src/bandersnatch/mirror.py

```diff
@@ -648,24 +652,49 @@ async def sync_release_files(self, package: Package) -> None:
         deferred_exception = None
         for release_file in package.release_files:
-            try:
-                downloaded_file = await self.download_file(
-                    release_file["url"],
-                    release_file["size"],
-                    datetime.datetime.fromisoformat(
-                        release_file["upload_time_iso_8601"].replace("Z", "+00:00")
-                    ),
-                    release_file["digests"]["sha256"],
-                )
-                if downloaded_file:
-                    downloaded_files.add(str(downloaded_file.relative_to(self.homedir)))
-            except Exception as e:
-                logger.exception(
-                    "Continuing to next file after error downloading: "
-                    f"{release_file['url']}"
-                )
-                if not deferred_exception:  # keep first exception
-                    deferred_exception = e
+            release_url = release_file["url"]
+            release_path = urlparse(release_url).path
+            if self.download_mirror and not self.download_mirror_no_fallback:
+                download_urls = [
+                    self.download_mirror + release_path,
+                    release_url,
+                ]
+            elif self.download_mirror and self.download_mirror_no_fallback:
+                download_urls = [
+                    self.download_mirror + release_path,
+                ]
+            else:
+                download_urls = [release_url]
+
+            for cnt, url in enumerate(download_urls):
+                try:
+                    downloaded_file = await self.download_file(
+                        url,
+                        release_file["size"],
+                        datetime.datetime.fromisoformat(
+                            release_file["upload_time_iso_8601"].replace("Z", "+00:00")
+                        ),
+                        release_file["digests"]["sha256"],
+                        urlpath=release_path,
+                    )
+                    if downloaded_file:
+                        downloaded_files.add(
+                            str(downloaded_file.relative_to(self.homedir))
+                        )
+                        break
+                except Exception as e:
+                    # Avoid flooding normal log messages with exception details
+                    if not len(download_urls) == (cnt + 1):
+                        logger.info(
+                            "Continuing to next file after error downloading: " f"{url}"
+                        )
+                    else:
+                        logger.exception(
+                            "Continuing to next file after error downloading: " f"{url}"
+                        )
+                    # keep previous exception, also ignore non-default urls
+                    if not deferred_exception and len(download_urls) == (cnt + 1):
+                        deferred_exception = e
```

Comment:
Follow up (or on this PR if you have cycles): this is complicated enough now maybe it warrants a function to explain in a doc string how we're looking up if we should use a mirror or fetch from URL in JSON (similar to what you put in the config).

Reply:
Will give this a shot soon.